mlos_bench/mlos_bench/storage/base_storage.py

```diff
@@ -0,0 +1,231 @@
+    """
+    An abstract interface between the benchmarking framework
+    and storage systems (e.g., SQLite or MLFLow).
+    """
+
+    def __init__(self, tunables: TunableGroups, service: Optional[Service], config: dict):
+        """
+        Create a new storage object.
+
+        Parameters
+        ----------
+        tunables : TunableGroups
+            Tunable parameters of the environment. We need them to validate the
+            configurations of merged-in experiments and restored/pending trials.
+        config : dict
+            Free-format key/value pairs of configuration parameters.
+        """
+        _LOG.debug("Storage config: %s", config)
+        self._tunables = tunables.copy()
+        self._service = service
+        self._config = config.copy()
+
+    @abstractmethod
+    def experiment(self, exp_id: str, trial_id: int, description: str, opt_target: str):
+        """
+        Create a new experiment in the storage.
+
+        Parameters
+        ----------
+        exp_id : str
+            Unique identifier of the experiment.
+        trial_id : int
+            Starting number of the trial.
+        description : str
+            Human-readable description of the experiment.
+        opt_target : str
+            Name of metric we're optimizing for.
+
+        Returns
+        -------
+        experiment : Storage.Experiment
+            An object that allows to update the storage with
+            the results of the experiment and related data.
+        """
+
+    class Experiment(metaclass=ABCMeta):
+        """
+        Base interface for storing the results of the experiment.
+        This class is instantiated in the `Storage.experiment()` method.
+        """
+
+        def __init__(self, tunables: TunableGroups, experiment_id: str,
+                     trial_id: int, description: str, opt_target: str):
+            self._tunables = tunables  # No need to copy, it's immutable
+            self._experiment_id = experiment_id
+            self._trial_id = trial_id
+            self._description = description
+            self._opt_target = opt_target
+            (self._git_repo, self._git_commit) = get_git_info()
+
+        def __enter__(self):
+            """
+            Enter the context of the experiment.
+            """
+            _LOG.debug("Starting experiment: %s", self)
+            return self
+
+        def __exit__(self, exc_type, exc_val, exc_tb):
+            """
+            End the context of the experiment.
+            """
+            if exc_val is None:
+                _LOG.debug("Finishing experiment: %s", self)
+            else:
+                _LOG.warning("Finishing experiment: %s", self,
+                             exc_info=(exc_type, exc_val, exc_tb))
+            return False  # Do not suppress exceptions
+
+        def __repr__(self) -> str:
+            return self._experiment_id
+
+        @abstractmethod
+        def merge(self, experiment_ids: List[str]):
```

Comment:
We'll have to add methods to discover the list of available ids as well.

Reply:
(future work I think)

Reply:
agree